FSW/controllers/detumbling_controller.py

```diff
@@ -0,0 +1,453 @@
+
+        rw_ang_vel = est_ctrl_states[Idx["X"]["RW_SPEED"]]
+
+        eci_pos = est_ctrl_states[Idx["X"]["ECI_POS"]]
+        eci_vel = est_ctrl_states[Idx["X"]["ECI_VEL"]]
+        orbit_vector = Re2b @ np.cross(eci_pos, eci_vel)
+        orbit_vector = orbit_vector / np.linalg.norm(orbit_vector)
+        if np.dot(orbit_vector,sun_vector) < 0:
+            orbit_vector = -orbit_vector
+        nadir_vector = Re2b @ (-eci_pos)
+        nadir_vector = nadir_vector / np.linalg.norm(nadir_vector)
+
+        # omega sat body frame ang vel
+        # omega_b gyro bias
+        # I_max - direction of axis of greatest inertia
+        
+        h = self.J @ angular_velocity 
+        h_norm = np.linalg.norm(h)
+        
+        # target quaternion from orbit_vector and nadir_vector
+        # nadir_cam_dir points t onadir_vector in target orientatio
+
+        u_mtb = np.zeros(3)
+        u_rw  = np.zeros(1)
+         
+        # angular momentum in a cone around the direction to be pointed
+        spin_stabilized = (h_norm <= (1.0 + np.deg2rad(15)) * self.h_tgt_norm) and \
+                          (np.arccos(np.clip(np.dot(self.I_min_direction, h / h_norm), -1.0, 1.0)) <= np.deg2rad(15))
+        # if angular velocity low enough, then its in nadir pointing and dont need to spin stabilize
+        spin_stabilized = spin_stabilized or (np.linalg.norm(angular_velocity) <= np.deg2rad(1))
+        # orbit_pointing = (np.linalg.norm(orbit_vector-(h/h_norm))<= np.deg2rad(10))
+        orbit_pointing = (np.linalg.norm(orbit_vector-self.G_rw_b.flatten())<= np.deg2rad(10))
+        magnetic_field_norm = np.linalg.norm(magnetic_field, ord=2)
+        
+        unit_magnetic_field = magnetic_field / (magnetic_field_norm ** 2)
+
+        if not spin_stabilized:
+            u_mtb = self.kdetumb * np.cross(unit_magnetic_field, self.ref_angular_velocity - angular_velocity).reshape(3,1)
+        elif not orbit_pointing: # pointing rw to orbit normal
+            Bhat = crossproduct(magnetic_field)
+            Bhat_pseudo_inv = Bhat.T / (magnetic_field_norm ** 2)
+
+            err_vec = np.hstack((-self.G_rw_b.flatten()-orbit_vector, self.ref_angular_velocity - angular_velocity))
+            u_mtb = Bhat_pseudo_inv @ self.k_mtb_att @ err_vec
+        else:
+            # nadir pointing controller
+            Bhat = crossproduct(magnetic_field)
+            """
+            U, S, V = np.linalg.svd(Bhat)
+            idx = np.where(S > 1e-6 * np.max(S))[0]
+            S_inv = np.diag([1/s if s > 1e-10 else 0 for s in S])
+            Bhat_pseudo_inv = V.T @ S_inv @ U.T"""
+            Bhat_pseudo_inv = Bhat.T / (magnetic_field_norm ** 2)
+            # 1. mtb nutation control
+            err_vec = np.hstack((-self.G_rw_b.flatten()-orbit_vector, angular_velocity))
+            u_mtb = Bhat_pseudo_inv @ self.k_mtb_att @ err_vec
+            u_mtb = np.clip(a=u_mtb, a_min=self.lbmtb, a_max=self.ubmtb)
```

Comment:
We really need to see how much time this takes. So many array ops will crash PyCubed

Reply:
this won't make it to FSW. This is our Model-In-the-loop code that only serves to prototype algorithms and to min trouble with sw dev. Only concern here is code being readable and easy to prototype new algos